core/src/KokkosExp_MDRangePolicy.hpp

```diff
@@ -357,112 +363,82 @@ struct MDRangePolicy : public Kokkos::Impl::PolicyTraits<Properties...> {
         m_prod_tile_dims(p.m_prod_tile_dims) {}
 
  private:
-  void init() {
-    // Host
-    if (true
+  // Host
+  template <typename ExecutionSpace>
+  void init(const ExecutionSpace&) {
+    const index_type max_threads = std::numeric_limits<int>::max();
+    const int max_tile_size      = std::numeric_limits<int>::max();
+    const int default_tile_size  = 2;
+    init_helper(max_threads, max_tile_size, default_tile_size);
+  }
+
 #if defined(KOKKOS_ENABLE_CUDA)
-        && !std::is_same<typename traits::execution_space, Kokkos::Cuda>::value
+  void init(const Kokkos::Cuda& space) {
+    const index_type max_threads =
+        space.impl_internal_space_instance()->m_maxThreadsPerSM;
+    const int max_tile_size     = 16;
+    const int default_tile_size = 2;
+    init_helper(max_threads, max_tile_size, default_tile_size);
+  }
 #endif
+
 #if defined(KOKKOS_ENABLE_HIP)
-        && !std::is_same<typename traits::execution_space,
-                         Kokkos::Experimental::HIP>::value
+  void init(const Kokkos::Experimental::HIP& space) {
+    const index_type max_threads =
+        space.impl_internal_space_instance()->m_maxThreadsPerSM;
+    const int max_tile_size     = 16;
+    const int default_tile_size = 4;
+    init_helper(max_threads, max_tile_size, default_tile_size);
+  }
 #endif
+
 #if defined(KOKKOS_ENABLE_SYCL)
-        && !std::is_same<typename traits::execution_space,
-                         Kokkos::Experimental::SYCL>::value
+  void init(const Kokkos::Experimental::SYCL& space) {
+    const index_type max_threads =
+        space.impl_internal_space_instance()->m_maxThreadsPerSM;
+    const int max_tile_size     = 16;
+    const int default_tile_size = 2;
+    init_helper(max_threads, max_tile_size, default_tile_size);
+  }
 #endif
-    ) {
-      index_type span;
-      for (int i = 0; i < rank; ++i) {
-        span = m_upper[i] - m_lower[i];
-        if (m_tile[i] <= 0) {
-          if (((int)inner_direction == (int)Right && (i < rank - 1)) ||
-              ((int)inner_direction == (int)Left && (i > 0))) {
-            m_tile[i] = 2;
-          } else {
-            m_tile[i] = (span == 0 ? 1 : span);
-          }
-        }
-        m_tile_end[i] =
-            static_cast<index_type>((span + m_tile[i] - 1) / m_tile[i]);
-        m_num_tiles *= m_tile_end[i];
-        m_prod_tile_dims *= m_tile[i];
-      }
+
+  void init_helper(index_type max_threads, int max_tile_size,
+                   int default_tile_size) {
+    index_type span;
```

Comment:
Given that we will eventually be using `std::span`, span is likely a bad name for this.  I know the original code does this, but if it is being changed anyway, we should pick a different name.

Reply:
Fixed.